src/Language/Marlowe/Runtime/AdjustTip.hs

```diff
@@ -0,0 +1,92 @@
+module Language.Marlowe.Runtime.AdjustTip
+  ( BuildChain(..)
+  , adjustTip
+  , processTip
+
+  -- re-exporting
+  , Seq
+  , (><)
+  , empty
+  , fromList
+  , toList
+  )
+  where
+
+import Data.Foldable ( toList )  -- Used in this module as the counterpart to Sequence.fromList
+import Data.Monoid ( First(..), getFirst )
+import Data.Sequence ( Seq, (><), drop, elemIndexL, empty, fromList, mapWithIndex, take )
+import Prelude hiding ( drop, take )
+
+
+{-
+before
+
+0..9  10..19  20..24..29  30..34,<24>,35..39
+ t1     t2        t3             chain
+
+after
+
+0..24,35..39 
+  newChain
+
+So what are the cases?
+
+1. We found the divergence point, added new tip and used rest of old chain.
+   Chain is complete
+2. We didn't find divergence point, added new tip to new chain and must keep
+   looking. Chain is NOT COMPLETE
+3. We ran out of new tip segments to search for and they've all been added.
+   Chain is complete. (This represents the first time we run with an empty
+   chain. Like server start-up)
+
+1 and 3 seem like the same thing? (Just newChain)
+-}
+
+
+data BuildChain a
+  = Building (Seq a)  -- Didn't find divergence point yet, need more tip segments
+  | BuildDone (Seq a)  -- Found divergence point or old chain exhausted
+  deriving (Eq, Show)
```

Comment:
This is really a complicated way of writing `(Bool, a)`. We can make a similar type to `Bool` to make it more semantic, but coupling it with `a` makes no sense to me.

**Edit:** Actually, a `Maybe` would be a better idea, see below

Reply:
I kind of like how the type constructor explain what the state of it is. What state does `(Bool, a)` say to a new developer?
The reason it's two different constructors is also that `Building` is carrying the cumulative new tip and `BuildDone` is carrying the new tail. Not the same contents.
The `a` was because I wrote this generic code that doesn't care what things it's merging-with-rollbacks. But we can use the actual type, sure, there's no value in the generality here.

Reply:
That is fine, but even then we could do either:
```haskell
data ChainState = Building
                | BuildDone
```
And then return `(ChainState, a)`, instead of adding `a` to both constructurs.
Or directly use `Maybe a` or can use your own version of `Maybe`:
```haskell
data ChainState = Building (Seq a)
                | BuildDone
```
But having it in every constructor makes no sense to me